File: data-plane/src/frontend/lib/ipam/subnet-mask-utils.ts
/**
 * Subnet Mask Utilities
 * 
 * Pure mathematical calculations for subnet mask reference information.
 * These values are deterministic and don't require backend API calls.
 */

export interface SubnetMaskInfo {
  bitmask: number;
  netmask: string;
  wildcard_mask: string;
  binary: string;
  subnets: number | string;
  hosts: number | string;
  subnet_bits: number;
  host_bits: number;
  is_ipv6?: boolean;
}

/**
 * Calculate subnet mask information for a given prefix length
 */
export function getSubnetMaskInfo(
  prefixLength: number,
  isIpv6: boolean = false
): SubnetMaskInfo {
  const maxPrefix = isIpv6 ? 128 : 32;

  if (prefixLength < 0 || prefixLength > maxPrefix) {
    throw new Error(`Prefix length must be between 0 and ${maxPrefix}`);
  }

  const hostBits = maxPrefix - prefixLength;
  const subnetBits = prefixLength;

  // Calculate number of hosts (2^host_bits)
  let totalHosts: number | string;
  if (hostBits > 64) {
    totalHosts = "Unlimited";
  } else {
    totalHosts = Math.pow(2, hostBits);
  }

  // For IPv4, exclude network and broadcast addresses
  let usableHosts: number | string;
  if (isIpv6) {
    usableHosts = totalHosts;
  } else {
    if (typeof totalHosts === "number") {
      usableHosts = Math.max(0, totalHosts - 2);
    } else {
      usableHosts = totalHosts;
    }
  }

  // Calculate number of subnets
  let subnets: number | string;
  if (isIpv6) {
    subnets = prefixLength < 128 ? Math.pow(2, 128 - prefixLength) : 1;
  } else {
    subnets = prefixLength < 32 ? Math.pow(2, 32 - prefixLength) : 1;
  }

  // Generate netmask
  let netmask: string;
  let wildcardMask: string;
  let binary: string;

  if (isIpv6) {
    // For IPv6, show prefix length notation (standard format)
    // e.g., /64, /48, etc.
    netmask = `/${prefixLength}`;
    
    // For IPv6, we can also show the mask in hex format
    // Calculate the mask value
    const maskParts: string[] = [];
    const fullHextets = Math.floor(prefixLength / 16);
    const partialBits = prefixLength % 16;
    
    for (let i = 0; i < fullHextets; i++) {
      maskParts.push("ffff");
    }
    
    if (partialBits > 0 && fullHextets < 8) {
      const maskValue = 0xffff << (16 - partialBits);
      maskParts.push(maskValue.toString(16).padStart(4, "0"));
    }
    
    // Fill remaining with zeros
    while (maskParts.length < 8) {
      maskParts.push("0000");
    }
    
    const hexMask = maskParts.join(":");
    wildcardMask = "Not applicable for IPv6";
    binary = hexMask; // Use hex for IPv6 (binary too long)
  } else {
    // For IPv4, calculate netmask from prefix length
    const maskValue = 0xffffffff << (32 - prefixLength);
    const maskBytes = [
      (maskValue >>> 24) & 0xff,
      (maskValue >>> 16) & 0xff,
      (maskValue >>> 8) & 0xff,
      maskValue & 0xff,
    ];
    netmask = maskBytes.join(".");

    // Calculate wildcard mask (inverse)
    const wildcardValue = 0xffffffff ^ maskValue;
    const wildcardBytes = [
      (wildcardValue >>> 24) & 0xff,
      (wildcardValue >>> 16) & 0xff,
      (wildcardValue >>> 8) & 0xff,
      wildcardValue & 0xff,
    ];
    wildcardMask = wildcardBytes.join(".");

    // Binary representation
    binary = maskBytes
      .map((b) => b.toString(2).padStart(8, "0"))
      .join(".");
  }

  return {
    bitmask: prefixLength,
    netmask,
    wildcard_mask: wildcardMask,
    binary,
    subnets,
    hosts: usableHosts,
    subnet_bits: subnetBits,
    host_bits: hostBits,
    is_ipv6: isIpv6,
  };
}

/**
 * Get all subnet masks for a given IP version
 */
export function getAllSubnetMasks(
  isIpv6: boolean = false,
  minPrefix?: number,
  maxPrefix?: number
): SubnetMaskInfo[] {
  const defaultMin = 0;
  const defaultMax = isIpv6 ? 128 : 32;

  const min = minPrefix ?? defaultMin;
  const max = maxPrefix ?? defaultMax;

  const masks: SubnetMaskInfo[] = [];
  // Generate from highest to lowest prefix length
  for (let prefix = max; prefix >= min; prefix--) {
    try {
      masks.push(getSubnetMaskInfo(prefix, isIpv6));
    } catch {
      continue;
    }
  }

  return masks;
}

/**
 * Get commonly used subnet masks
 */
export function getCommonSubnetMasks(isIpv6: boolean = false): SubnetMaskInfo[] {
  if (isIpv6) {
    // Common IPv6 prefix lengths
    const commonPrefixes = [
      128, 127, 126, 125, 124, 120, 112, 108, 104, 96, 80, 64, 48, 32, 24, 16, 8, 0,
    ];
    return commonPrefixes
      .map((prefix) => {
        try {
          return getSubnetMaskInfo(prefix, true);
        } catch {
          return null;
        }
      })
      .filter((mask): mask is SubnetMaskInfo => mask !== null);
  } else {
    // Common IPv4 prefix lengths (from /32 to /8)
    const commonPrefixes = [
      32, 31, 30, 29, 28, 27, 26, 25, 24, 23, 22, 21, 20, 19, 18, 17, 16, 15, 14,
      13, 12, 11, 10, 9, 8, 7, 6, 5, 4, 3, 2, 1, 0,
    ];
    return commonPrefixes
      .map((prefix) => {
        try {
          return getSubnetMaskInfo(prefix, false);
        } catch {
          return null;
        }
      })
      .filter((mask): mask is SubnetMaskInfo => mask !== null);
  }
}

File: data-plane/src/backend/ipam/services/subnet_mask_utils.py
"""
Subnet Mask Utilities

Provides functions for calculating and displaying subnet mask information.
"""

import ipaddress
from typing import Dict, List


def get_subnet_mask_info(prefix_length: int, is_ipv6: bool = False) -> Dict:
    """
    Get detailed information about a subnet mask for a given prefix length.

    Args:
        prefix_length: CIDR prefix length (0-32 for IPv4, 0-128 for IPv6)
        is_ipv6: Whether this is an IPv6 subnet mask

    Returns:
        Dictionary containing subnet mask information:
        - bitmask: Prefix length
        - netmask: Subnet mask in dotted decimal notation (IPv4) or hex (IPv6)
        - wildcard_mask: Wildcard mask (inverse of netmask)
        - binary: Binary representation
        - subnets: Number of possible subnets
        - hosts: Number of usable hosts
        - subnet_bits: Number of subnet bits
        - host_bits: Number of host bits
    """
    if is_ipv6:
        max_prefix = 128
    else:
        max_prefix = 32

    if prefix_length < 0 or prefix_length > max_prefix:
        raise ValueError(f"Prefix length must be between 0 and {max_prefix}")

    # Calculate host bits
    host_bits = max_prefix - prefix_length
    subnet_bits = prefix_length

    # Calculate number of hosts (2^host_bits)
    total_hosts = 2**host_bits if host_bits <= 64 else float("inf")

    # For IPv4, exclude network and broadcast addresses
    if is_ipv6:
        usable_hosts = total_hosts
    else:
        usable_hosts = max(0, total_hosts - 2) if total_hosts != float("inf") else total_hosts

    # Calculate number of subnets (2^subnet_bits)
    # For reference purposes, we show how many /32 subnets can fit in this prefix
    if is_ipv6:
        subnets = 2 ** (128 - prefix_length) if prefix_length < 128 else 1
    else:
        subnets = 2 ** (32 - prefix_length) if prefix_length < 32 else 1

    # Generate netmask
    if is_ipv6:
        # For IPv6, create a network with the prefix length
        net = ipaddress.IPv6Network(f"::/{prefix_length}", strict=False)
        netmask = str(net.netmask)
        wildcard_mask = "Not applicable for IPv6"
        binary = _ipv6_to_binary(net.netmask)
    else:
        # For IPv4, create a network with the prefix length
        net = ipaddress.IPv4Network(f"0.0.0.0/{prefix_length}", strict=False)
        netmask = str(net.netmask)
        # Calculate wildcard mask (inverse of netmask)
        wildcard_int = 0xFFFFFFFF ^ int(net.netmask)
        wildcard_mask = str(ipaddress.IPv4Address(wildcard_int))
        binary = _ipv4_to_binary(net.netmask)

    return {
        "bitmask": prefix_length,
        "netmask": netmask,
        "wildcard_mask": wildcard_mask,
        "binary": binary,
        "subnets": subnets,
        "hosts": usable_hosts if usable_hosts != float("inf") else "Unlimited",
        "subnet_bits": subnet_bits,
        "host_bits": host_bits,
        "is_ipv6": is_ipv6,
    }


def _ipv4_to_binary(ip: ipaddress.IPv4Address) -> str:
    """Convert IPv4 address to binary string with dots."""
    binary_str = format(int(ip), "032b")
    return ".".join([binary_str[i : i + 8] for i in range(0, 32, 8)])


def _ipv6_to_binary(ip: ipaddress.IPv6Address) -> str:
    """Convert IPv6 address to binary string with colons (hex format for readability)."""
    # For IPv6, show hex representation instead of full binary (too long)
    hex_str = format(int(ip), "032x")
    return ":".join([hex_str[i : i + 4] for i in range(0, 32, 4)])


def get_all_subnet_masks(
    is_ipv6: bool = False, min_prefix: int = None, max_prefix: int = None
) -> List[Dict]:
    """
    Get subnet mask information for all prefix lengths.

    Args:
        is_ipv6: Whether to generate IPv6 subnet masks
        min_prefix: Minimum prefix length (default: 0 for IPv4, 0 for IPv6)
        max_prefix: Maximum prefix length (default: 32 for IPv4, 128 for IPv6)

    Returns:
        List of dictionaries containing subnet mask information
    """
    if is_ipv6:
        default_min = 0
        default_max = 128
    else:
        default_min = 0
        default_max = 32

    min_prefix = min_prefix if min_prefix is not None else default_min
    max_prefix = max_prefix if max_prefix is not None else default_max

    masks = []
    # Generate from highest to lowest prefix length (most specific to least specific)
    for prefix in range(max_prefix, min_prefix - 1, -1):
        try:
            mask_info = get_subnet_mask_info(prefix, is_ipv6)
            masks.append(mask_info)
        except ValueError:
            continue

    return masks


def get_common_subnet_masks(is_ipv6: bool = False) -> List[Dict]:
    """
    Get subnet mask information for commonly used prefix lengths.

    Args:
        is_ipv6: Whether to generate IPv6 subnet masks

    Returns:
        List of dictionaries containing subnet mask information for common prefix lengths
    """
    if is_ipv6:
        # Common IPv6 prefix lengths
        common_prefixes = [
            128,
            127,
            126,
            125,
            124,
            120,
            112,
            108,
            104,
            96,
            80,
            64,
            48,
            32,
            24,
            16,
            8,
            0,
        ]
    else:
        # Common IPv4 prefix lengths (from /32 to /8)
        common_prefixes = [
            32,
            31,
            30,
            29,
            28,
            27,
            26,
            25,
            24,
            23,
            22,
            21,
            20,
            19,
            18,
            17,
            16,
            15,
            14,
            13,
            12,
            11,
            10,
            9,
            8,
            7,
            6,
            5,
            4,
            3,
            2,
            1,
            0,
        ]

    masks = []
    for prefix in common_prefixes:
        try:
            mask_info = get_subnet_mask_info(prefix, is_ipv6)
            masks.append(mask_info)
        except ValueError:
            continue

    return masks
